lib/spack/spack/package.py

```diff
@@ -973,6 +973,80 @@ def url_for_version(self, version):
         return spack.url.substitute_version(
             default_url, self.url_version(version))
 
+    def all_urls_for_version(self, version):
+        """Returns all URLs derived from version_urls(), url, urls, and
+        list_url (if it contains a version) in a package in that order.
+
+        version: class Version
+            The version for which a URL is sought.
+        use_list_url: bool
+            Whether to use list_url over url field
+
+        See Class Version (version.py)
+        """
+        if not isinstance(version, Version):
+            version = Version(version)
+
+        urls = []
+
+        # If we have a specific URL for this version, don't extrapolate.
+        version_urls = self.version_urls()
+        if version in version_urls:
+            urls.append(version_urls[version])
+
+        # if there is a custom url_for_version, use it
+        u = self.url_for_version(version)
```

Comment:
Regarding:

> This turns out to be harder than I would have thought. Since url_for_version is the customization point for packages, I need to call it in all_urls_for_version. I tried to detect when it's customized and only call it then, but that is proving to be unreliable. I'll be happy to do the rename, but getting the singular to use the multiple is not straightforward.

We could:

* Move all of this logic to a new function
* Add a `_custom_url_function` parameter to that function
* `all_urls_for_version` sets `_custom_url_function=url_for_version`
*  `url_for_version` can set `_custom_url_function=None`

Reply:
Assuming my updated test for overloading, this should be in and working now.